### ISSUE TITLE
JDBC Input and Output format for Stratosphere

### ISSUE THREAD

**Issue by emrehasan** [stratosphere/stratosphere#261]
Hi,

I would like to contribute to Stratosphere too. On your page https://github.com/stratosphere/stratosphere/wiki/Starter-Jobs i found the task 'JDBC Input and Output format for Stratosphere'  and would like to get more information on that.

**Comment by rmetzger** [comment]
Hi,

Cool, we are very happy that you want to contribute to the project!

I would recommend you to start with implementing the JDBC Input Format. Create a class `JDBCInputFormat` that extends from `eu.stratosphere.pact.common.io.GenericInputFormat`.  The overall idea of the InputFormat is being able to read from a database with a JDBC driver.
So I would recommend you to set up a MySQL database (or mariadb, or anything that's open source and has JDBC support) and populate it with some data.
Then, write a little test stratosphere job that reads the result of a query into a CSV file.
Implement your `JDBCInputFormat` until your job runs successfully. 

You have to override some methods for the input format to work:
- `nextRecord()` is called by the framework to get the nextRecord from the input. So in your case, this method should return the next row from the query result.
- `reachedEnd()` the loop that calls `nextRecord()` checks with `reachedEnd()` if there are records left.
- The GenericInputFormat has a mechanism to create "input splits". This is the way Stratosphere reads files in parallel. I would recommend you to circumvent this mechanism at first and only use one input split for the input format. (so the input format does not run in parallel, but that's okay since I would not recommend loading "big data" scale data)
- I would suggest to create a constructor for the `JDBCInputFormat` with two arguments, one for the JDBC connection and one for a SQL Query.

Please ask here if you have questions or if you need help with anything. (I can offer you to review your code. We can also setup a Skype call (german or english) if you need further assistance)

Please also report any issues, inconveniences or missing information. This is also very valuable information for us, since we want to make the start with our system as easy as possible.

**Comment by emrehasan** [comment]
Should the PactRecord be filled with just Strings like it is done in TextInputFormat or should the original datatypes be considered? If yes which sql-types shall be considered?

**Comment by rmetzger** [comment]
It would be nice to map the SQL types to the respective Pact\* types, but I think the most basic types such as int, long, string, bool etc. should be sufficient for a first approach.

If you have types that are absolutely not mappable, map them to string or throw an exception.

**Comment by emrehasan** [comment]
How should the test be modeled? Shall it just be a test so that i am convinced that the code of my partner and me works or should there be a real test-unit? Therefore we would need an extern configuration-file like in /test/resources so that every dev can run the test after setting up his own database-connections too.

**Comment by rmetzger** [comment]
I would create a new class called `JDBCInputExample` that shows a whole Stratosphere job (as I described before).
Place the schema of the table you've used either into a comment inside the example job or into the `resources` directory. Ideally, the example schema should contain different data types to test the conversion.

A unit test is not required for a first approach, I would first focus on having a running example. In the long term however, we could think about implementing a unit test using a derby database (which runs out of java and in-memory). Having a unit test is definitively a good idea to ensure the stability of the input format.
Our unit tests are designed to run without any manual configuration.

**Comment by zentol** [comment]
Hello, I'm the partner of emrehasan, we work together on the issue.

We were able to retrieve data from a locally setup mysql database. 
connection establishment, nextRecord() and reachedEnd() all appear to work as they should.
(no inputsplits as you suggested)

Since getting stratosphere running on windows turned out problematic, we instead put the jdbcinputexample into a junit test (without the csv part, just assertions tailored for the data). i hope that's alright.

**Comment by uce** [comment]
Hey @zentol,

can you also please provide more information on what is not working on Windows?

**Comment by rmetzger** [comment]
Hi,

we recently added code to fix many bugs on windows (https://github.com/stratosphere/stratosphere/commit/1edf660e85a19772771f4aa50407f7e5dd860385). Did you merge the most recent changes of our main repo into your code line.

If you still have the problems after the merge, please report all bugs you encounter when using Stratopshere on Windows.

**Comment by zentol** [comment]
it wasnt a problem with stratosphere itself, setting up cygwin, curl and maven properly was the issue. instead of spending time on that we instead opted for using junit to get working code asap.

i actually managed to run the wordcount job, so we're going to create a proper example now.

**Comment by emrehasan** [comment]
Currently i am developing the next version of the test-unit:
Using Apache Derby was supposed to replace the example later and to satisfy correct running code by just automated tests. It should be implemented with all necessary condition-checkings within the next week.

**Comment by rmetzger** [comment]
Hi guys,
@zentol: Okay, I understand. I like your pragmatism, don't get stuck by those issues. I thought Stratosphere can run on Windows (I think you have to instruct git to change the line endings or so, but then it should work). We have a Stratosphere developer who will probably switch to Windows soon, he'll figure it out and make Stratosphere compatible. (If you have found bugs that only occur on windows, tell us! (Bugs reports are a really important contribution for us).
I'm looking forward to see your example ;)

@emrehasan: Great! Its really good that you provide Unit-tests!

**Comment by zentol** [comment]
I just finished the example. While working on it i took a look at other examples/input formats/way-too-many-other- classes and noticed a few oddities.

All other examples use InputFormats that have a parameter-less constructor. They create a datasource, and subsequently create the configuration using datasource.setParameter().

Our constructor on the other hand expects a configuration(to set the connection parameters) and a string(query), which means the configuration has to be created before the datasource.

Should we change our parameter and example accordingly? i actually have two versions right now for each case, it would just be a matter of deciding which one to use.

**Comment by rmetzger** [comment]
Hi,

Your input format must have a parameter free constructor. The reason for this is, that the input format is created in the following situations:
-  When Stratosphere starts reading data (from files or databases), it first has to create the InputSplits. Therefore, the InputFormat is called by the user (here, you can pass arguments using the constructor (this feature is not often used in the examples because we support it only recently https://github.com/stratosphere/stratosphere/pull/73)
- Each time a InputSplit starts reading its assigned split (usually a part of a file) the InputFormat is opened using the argument free constructor. 

So I think it is okay for you to get the query using the constructor. Just make sure that its available in the InputSplit. (As I said before, write the InputFormat so that there is only one InputSplit).

I'll take a look into your code, I assume its in https://github.com/emrehasan/stratosphere.

**Comment by rmetzger** [comment]
Okay, looking at the example (https://github.com/emrehasan/stratosphere/blob/c64e0694e0a042e6fef952e643a8fd6e00989a71/pact/pact-examples/src/main/java/eu/stratosphere/pact/example/jdbcinput/JDBCInputExample.java) I would suggest to use the version with two arguments (config and query).
Could you also add a constructor where the user can enter the jdbc connection as a string. I think they look like "mysql://user@password:db" or so? This way, one can define connection and query in one line, which would be quite nice. (And I assume most Java developers know how JDBC works)

Concerning the configuration: If you code currently supports both (Config object as argument) and the setParameter() variant, keep both. I think it does not hurt anybody! 
If not, I like the config object to the ctor variant more.

In general, your example looks very nice! I really like it!

**Comment by zentol** [comment]
pretty sure i can create a workaround so that both work. the constructor you requested is already done.

Update:

so we got 3 constructors now:

1:()
2:(url,query)
3:(configuration,query)

1 expects that all parameters are passed using datasource.setParameter, including the query
2 expects a complete url that can be passed directly into the getConnection call
3 expects all parameters to be passed via the configuration it receives

2 and 3 ignore all parameters sent via datasource.setParameter.

I've added an example for each case.

**Comment by rmetzger** [comment]
Hi,

looks perfect. Thanks!

**Comment by zentol** [comment]
We are currently finalizing our code, are there any guidelines for comments/formatting we should follow?

**Comment by rmetzger** [comment]
Hi,

Thats a good question. I'd suggest that I'll review your code for that.
I already looked at it a few days ago and I liked it.

Sent from my iPhone

On 20.11.2013, at 10:57, zentol notifications@github.com wrote:

We are currently finalizing our code, are there any guidelines for
comments/formatting we should follow?

—
Reply to this email directly or view it on
GitHubhttps://github.com/stratosphere/stratosphere/issues/261#issuecomment-28875174
.

**Comment by StephanEwen** [comment]
The input format is done (4d73ef004199fb933bf4669da63e4254ad115c3f). We are keeping the issue open for the output format.

**Comment by zentol** [comment]
It's been way too long since we wrote something in here, so here's an update!

I think we are done with the jdbcoutputformat, or rather i wouldn't know what the next step would be.

We have a test for the basic functionality, as in writing data. exceptions aren't tested.
The test reads data from a db, writes it to a second, then reads from the second and matches it.

an issue we are having is that the example wont run, and the exceptions thrown arent particularly helpful to me :/

ERROR client.JobClient: ERROR: eu.stratosphere.nephele.executiongraph.GraphConversionException: java.lang.NoSuchMethodError: eu.stratosphere.pact.common.util.InstantiationUtil.readObjectFormConfig(Leu/stratosphere/nephele/configuration/Configuration;Ljava/lang/String;Ljava/lang/ClassLoader;)Ljava/lang/Object;
    at eu.stratosphere.pact.runtime.task.util.TaskConfig.getStubWrapper(TaskConfig.java:259)
...

ERROR eu.stratosphere.nephele.client.JobClient                      - ERROR: eu.stratosphere.nephele.executiongraph.GraphConversionException: java.lang.NoSuchMethodError: eu.stratosphere.pact.common.util.InstantiationUtil.readObjectFormConfig(Leu/stratosphere/nephele/configuration/Configuration;Ljava/lang/String;Ljava/lang/ClassLoader;)Ljava/lang/Object;
    at eu.stratosphere.pact.runtime.task.util.TaskConfig.getStubWrapper(TaskConfig.java:259)
...

Exception in thread "main" eu.stratosphere.nephele.client.JobExecutionException: eu.stratosphere.nephele.executiongraph.GraphConversionException: java.lang.NoSuchMethodError: eu.stratosphere.pact.common.util.InstantiationUtil.readObjectFormConfig(Leu/stratosphere/nephele/configuration/Configuration;Ljava/lang/String;Ljava/lang/ClassLoader;)Ljava/lang/Object;
    at eu.stratosphere.pact.runtime.task.util.TaskConfig.getStubWrapper(TaskConfig.java:259)
...

**Comment by fhueske** [comment]
@zentol, @emrehasan, @rmetzger: Great work! 
Anything left to do or is this issue ready to be closed?
